### ISSUE TITLE
Debugger keeps stoping at Disabled Breakpoints when debugging in Chrome

### ISSUE THREAD

**Issue by awsms** [Dart-Code/Dart-Code#5144]
**Describe the bug**
I have set a bunch of breakpoints for my Flutter project. When I click "Toggle Active Breakpoints", or manually uncheck breakpoints, their state will be ignored, and the debugger will still stops the code at the said disabled breakpoints.

**To Reproduce**
1. Add a breakpoint
2. Run the project in debug mode
3. Toggle the breakpoint so it's inactive in VSCode
4. Code keeps stopping at said breakpoint, untill I kill the app & re-run the project in debug mode

**Expected behavior**
The debugger should ignore inactive breakpoints.

**Screenshots**
![image](https://github.com/Dart-Code/Dart-Code/assets/48278661/25f8e2e5-96fd-4dcd-b893-3e75b1477e6f)

**Please complete the following information:**

You can run the **Dart: Collect Diagnostic Information** command from the VS Code command palette (`F1`) to easily capture this information or provide it manually.

 - Operating System and version: EndeavourOS
 - VS Code version: VSCodium 1.90.1
 - Dart extension version: v3.90.0
 - Dart/Flutter SDK version: 3.4.1 (stable) / 3.22.1 (stable)

**Comment by awsms** [comment]
> I'm not able to reproduce this. Could you confirm which kind of device you're running on (web, mobile, desktop?), and are you able to reproduce this in a newly created application so I can test with exactly the same project?
> 
> Thanks!

Indeed, I just tried on the counter project (`flutter create count`) and I wasn't able to reproduce it at all.
I'm only experiencing it while debugging my project in Web, the Linux build stops at the breakpoints it's supposed to.
Is it possible the browser is the one keeping the breakpoints in memory until I restart the project?
By the way, I'm using thorium for debugging (https://github.com/Alex313031/thorium), in case that's relevant.

**Comment by DanTup** [comment]
> Is it possible the browser is the one keeping the breakpoints in memory until I restart the project?

There were some differences with how breakpoints were persisted in the browser, but toggling them off should still work (and it sounds like it does for you in the counter app).

How reliably can you reproduce this in your project? Is it only after a Hot Reload/Hot Restart, or can you repro without those?

Can you reproduce it in any public projects that I could test with?

**Comment by awsms** [comment]
Yes! I was just able to reproduce this bug with https://github.com/lucavenir/go_router_riverpod

**Comment by DanTup** [comment]
@awsms do you have exact steps to repro? I've got that project cloned but I'm hitting another issue running it in web which I'm looking into, but knowing where you're adding breakpoints/removing them and then triggering them would help me track it down once I've solved this. Thanks!

**Comment by DanTup** [comment]
I got the app running, and I tried the following:

- Add a breakpoint in `lib\state\auth_controller.dart` in `login()`
- Click the login button
- Breakpoint triggered
- Resume debugger
- Remove the breakpoint
- Click the logout button
- Click the login button
- Breakpoint was _not_ hit

Can you clarify your exact sequence of actions? Thanks!

**Comment by awsms** [comment]
Mh interesting, I'm still having this:
![image](https://github.com/Dart-Code/Dart-Code/assets/48278661/5fa20733-fb19-4f6c-bd50-ce09eca317c3)

- Add a breakpoint to auth_controller:dart:l51
- Login, then logout
- Breakpoint triggered
- Continue
- Disable breakpoint temporary
- Login, logout again
- Breakpoint still here

Then I redid the same last steps, but completely removed the breakpoint, and it's still here:
![image](https://github.com/Dart-Code/Dart-Code/assets/48278661/38cfb432-75c4-43cb-b84e-2f7143186eb9)

**Comment by DanTup** [comment]
Thanks - unfortunately I still can't trigger with those exact steps.

Could you:

1. confirm you're running using the Chrome device (not Edge or web-server or something)
2. capture verbose logs for the session
  1. run the **Dart: Capture Debugging Logs** command from the command palette **before starting the debug session**
  2. follow the steps above to reproduce the issue (first triggering the breakpoint, then removing it, then triggering it again)
  3. click **cancel** on the logging notification to stop logging and open the log file
  4. attach the log here (or paste into a gist and then the link here) - be sure to remove anything sensitive first (but since this is a public project, that should be limited to things like paths)

Thanks!

**Comment by stan-at-work** [comment]
+1 this is a big problem, Break points in general are broken in with extension

**Comment by DanTup** [comment]
@stan-at-work please don't add "+1" comments to lots of different issues, it causes lots of notifications and doesn't add anything.

If you're having issues with breakpoints, please describe the issues you're seeing in detail. This issue I could not reproduce and I left instructions above for capturing logs. If you're seeing this same issue, then please follow those instructions to produce a log file.

The fastest way to getting any issue solved is to help me reproduce it.

**Comment by awsms** [comment]
I've just updated VSCodium to 1.90.2.24171, and I can't reproduce this bug right now (in the go_router_riverpod project). But I'll be sure to report here again if I encounter this one more time, later this day.

Btw, I also experience another bug with the Dart debugger+chrome web, sometimes the browser will lock itself to a breakpoint, and despite hot reloading/refreshing the browser/doing anything, nothing happens. The only thing I can see is the "Flutter: Hot restarting..." notifications in VSCode stacking up at each hot reload. Is this a known bug/issue?
Otherwise I could also try reproducing and generate debug logs for this one + opening another issue.

**Comment by DanTup** [comment]
@awsms I wouldn't expect the VS Code / VSCodium version to make much difference here so it might just be inconsistent. If you're able to repro again please do share logs.


> Btw, I also experience another bug with the Dart debugger+chrome web, sometimes the browser will lock itself to a breakpoint, and despite hot reloading/refreshing the browser/doing anything, nothing happens. The only thing I can see is the "Flutter: Hot restarting..." notifications in VSCode stacking up at each hot reload. Is this a known bug/issue?

Could be https://github.com/Dart-Code/Dart-Code/issues/4980 (which I think is a Flutter issue - see https://github.com/flutter/flutter/issues/143371).


>  I also have this bug since I started learning Flutter, whenever I debug a build for Linux, if a key is pressed once the app is built, the whole keymaps will be buggy (can't use Tab, Backspace etc), and my debugger logs are filled with "A KeyDownEvent is dispatched, but the state shows that the physical key is already pressed. If this occurs in real application, please report this bug to Flutter.". Is this a known bug as well?

There are a few issues in Flutter for this error (https://github.com/flutter/flutter/issues?q=is%3Aissue+is%3Aopen+%22A+KeyDownEvent+is+dispatched%22+) but I don't know if they're exactly the same - I would review them and if you think yours isn't the same, open an issue in flutter/flutter with steps to repro. Thanks!

**Comment by awsms** [comment]
Hi! So I was able to reproduce the bug, with logs for it. All I had to do was hot refresh the go_router_riverpod project a bunch of time, and voilà.
https://krakenfiles.com/view/RFoJX9czUw/file.html

**Comment by DanTup** [comment]
@awsms that link tries to serve me malware if I click download (either exes or trying to set up browser notifications). You'll need to share the log in another way.

**Comment by awsms** [comment]
> @awsms that link tries to serve me malware if I click download (either exes or trying to set up browser notifications). You'll need to share the log in another way.

Mh I've been using this filehoster for years without issue (I've an adblocker though). Github didn't allow .7z files, which have better compression levels for text files. I'll upload a zip here:
[Dart-Code-Log-2024-06-05 13-41-18.zip](https://github.com/user-attachments/files/16129828/Dart-Code-Log-2024-06-05.13-41-18.zip)

**Comment by DanTup** [comment]
For me, every click goes to a dodgy looking domain and tries to give files like "OperaSetup.exe", "Dart-Code-Log-2024-06-05 13-41-18.exe" or triggers notifications in Chrome about allowing push notifications. I don't know if it's coming from the site itself of some dodgy ads they're allowing, but either way it doesn't seem very trustworthy to me 😔

Thanks for re-uploading - I have the file now and will do some digging. Thanks!

**Comment by DanTup** [comment]
Thanks for the log - I think this is the same issue that's come up here:

- https://github.com/dart-lang/webdev/issues/1529
- https://github.com/dart-lang/webdev/issues/2257

A fix was landed in https://github.com/dart-lang/webdev/pull/2371 which according to https://github.com/dart-lang/webdev/blob/master/dwds/CHANGELOG.md#2400 was included in DWDS 24.0.

The current stable Flutter version (3.22.2) is only using DWDS 23.3.0:

https://github.com/flutter/flutter/blob/3.22.2/packages/flutter_tools/pubspec.yaml#L17

The current _beta_ is on DWDS 24.0 though, so I would expect the issue is resolved and can be tested there. If you can still reproduce in the current beta, please let me know (ideally, a log file and a screencapture or detailed description of your actions would help). Thanks!

**Comment by Sten435** [comment]
> Thanks for the log - I think this is the same issue that's come up here:
> 
> 
> 
> - https://github.com/dart-lang/webdev/issues/1529
> 
> - https://github.com/dart-lang/webdev/issues/2257
> 
> 
> 
> A fix was landed in https://github.com/dart-lang/webdev/pull/2371 which according to https://github.com/dart-lang/webdev/blob/master/dwds/CHANGELOG.md#2400 was included in DWDS 24.0.
> 
> 
> 
> The current stable Flutter version (3.22.2) is only using DWDS 23.3.0:
> 
> 
> 
> https://github.com/flutter/flutter/blob/3.22.2/packages/flutter_tools/pubspec.yaml#L17
> 
> 
> 
> The current _beta_ is on DWDS 24.0 though, so I would expect the issue is resolved and can be tested there. If you can still reproduce in the current beta, please let me know (ideally, a log file and a screencapture or detailed description of your actions would help). Thanks!

When will this go live (v24)

**Comment by DanTup** [comment]
> When will this go live (v24)

This is down to when Flutter releases, something I'm not involved in. Flutter usually releases [approximately quarterly](https://github.com/flutter/flutter/blob/master/docs/roadmap/Roadmap.md#releases).
